### PR TITLE
Delivering All Outputs

### PR DESCRIPTION
**Pull request by andela-pessien**
#### What does this PR do?
It contains the outputs for:
1. Automating Tasks
2. Checking Repository Health
3. Delivering Your Product
4. Getting Peer Reviews
5. Reflecting on the Learning Experience
6. Setting up Continuous Integration
7. Setting up Linters
8. Setting up Software Deployment Environments
9. Setting up Your Repository
10. Solving the Problem
11. Using Design Patterns
12. Writing HTML/JS/CSS Outputs
13. Writing Tests

#### Description of Task to be completed?
Prepare and submit outputs for each context for the checkpoint.
1. **Automating Tasks**: Deliver a screencast showing gulp watching and refreshing your browser when changes are made to your HTML, CSS & JS files.
2. **Checking Repository Health**: Deliver a screenshot of your code coverage and GPA badges in your repo's README file.
3. **Delivering Your Product**: Deliver release notes for your final product in your product's wiki.
4. **Getting Peer Reviews**: Deliver a Pull Request showing at least two Peer Code Reviews.
5. **Reflecting on the Learning Experience**: Deliver a post reflecting on your learning experience while producing these outputs.
6. **Setting up Continuous Integration**: Deliver a link to a successful TravisCI build of your development branch (it should include test runs).
7. **Setting up Linters**: Deliver two screenshots of ESLint Airbnb style violations in the terminal and HoundCI comment on a PR.
8. **Setting up Software Deployment Environments**: Deliver live URLs to a review app, your staging & production apps.
9. **Setting up Your Repository**: Deliver the repository for your proposed project and at least one Pull Request.
10. **Solving the Problem**: Deliver the file containing the implementation including all relevant dependencies. (e.g. InvertedIndex class).
11. **Using Design Patterns**: Deliver a Pull Request showing two Design Patterns (suggestions: Decorator, Factory, Observer, Prototype...) or a detailed explanation of Design Patterns used while `solving-the-problem`.
12. **Writing HTML/JS/CSS Outputs**: Deliver a folder containing the template for the proposed frontend of your project.
13. **Writing Tests**: Deliver your test files for the Inverted Index class and all associated dependencies.

#### How should this be manually tested?
Links to the outputs can be found in the `output` folder inside the folders for each context.
1. **Automating Tasks**: Watch the YouTube video linked in `06-building-a-program-start-to-finish/automating-tasks/output/README.md`.
2. **Checking Repository Health**: View the screenshot linked in `06-building-a-program-start-to-finish/checking-repository-health/output/README.md`.
3. **Delivering Your Product**: Follow the link to the release notes in `06-building-a-program-start-to-finish/delivering-your-product/output/README.md`.
4. **Getting Peer Reviews**: Follow the link to the PR in `06-building-a-program-start-to-finish/getting-peer-reviews/output/README.md`.
5. **Reflecting on the Learning Experience**: Follow the link to the Medium post in `06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md`.
6. **Setting up Continuous Integration**: Follow the link to the Travis build in `06-building-a-program-start-to-finish/setting-up-continuous-integration/output/README.md`.
7. **Setting up Linters**: View the screenshots linked in `06-building-a-program-start-to-finish/setting-up-linters/output/README.md`.
8. **Setting up Software Deployment Environments**: Launch and try out the apps linked in `06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md`.
9. **Setting up Your Repository**: Follow the links to the repo and PR in `06-building-a-program-start-to-finish/setting-up-your-repository/output/README.md`.
10. **Solving the Problem**: View the InvertedIndex class file at `06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js`.
11. **Using Design Patterns**: Read the explanation in `06-building-a-program-start-to-finish/using-design-patterns/output/README.md`.
12. **Writing HTML/JS/CSS Outputs**: View the frontend template files in `06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/public/`.
13. **Writing Tests**: View the spec files at `06-building-a-program-start-to-finish/writing-tests/output/test/`.

#### Any background context you want to provide?
None right now

#### What are the relevant pivotal tracker stories?
N/A

#### Questions:
None right now

### CODE REVIEW

**Comment by andela-moe** [comment]
Your title and description content for this PR are not properly done. Please meet your team mates for more clarity.

File: 06-building-a-program-start-to-finish/automating-tasks/output/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: Gulp
+- **Deliver**: a screencast showing gulp watching and refreshing your browser when changes are made to your HTML, CSS & JS files
+- **With**: *independently*
+
+[Gulp/Browsersync screencast](http://www.youtube.com/watch?v=jZcnAT1L700)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/automating-tasks/output/README.md:6]: Nice

File: 06-building-a-program-start-to-finish/checking-repository-health/output/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: `Code Climate` and `Coveralls`
+- **Deliver**: a screenshot of your code coverage and GPA badges in your repo's README file
+- **With**: *independently*
+
+[README with badges](https://www.dropbox.com/s/jciag9o12xa9yej/badges.png?dl=0)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/checking-repository-health/output/README.md:6]: Cool

File: 06-building-a-program-start-to-finish/getting-peer-reviews/output/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: GitHub
+- **Deliver**: a Pull Request showing at least two Peer Code Reviews
+- **With**: *independently*
+
+[Pull Request with peer reviews](https://github.com/andela-pessien/inverted-index/pull/1)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/getting-peer-reviews/output/README.md:6]: Ok. Were you able to implement the feedback you received?

File: 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: TravisCI
+- **Deliver**: a link to a successful TravisCI build of your `development` branch (it should include test runs)
+- **With**: *independently*
+
+[Successful Travis build](https://travis-ci.org/andela-pessien/inverted-index/builds/214462312)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md:6]: Cool

File: 06-building-a-program-start-to-finish/setting-up-linters/output/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+- **Using**: Terminal & GitHub
+- **Deliver**: two screenshots of ESLint [Airbnb](https://github.com/airbnb/javascript)  style violations in the terminal and [HoundCI](https://houndci.com/) comment on a PR
+- **With**: *independently*
+
+[ESLint Screenshot 1](https://www.dropbox.com/s/0s18bxl3g5biuup/eslint1.png?dl=0)
+
+[ESLint Screenshot 2](https://www.dropbox.com/s/eu1u931kp0ysj0v/eslint2.png?dl=0)
+
+[HoundCI Screenshot](https://www.dropbox.com/s/vyd18ckj0461wox/houndci.png?dl=0)
+[HoundCI Review](https://github.com/andela-pessien/inverted-index/pull/1#pullrequestreview-28628948)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-linters/output/README.md:11]: Cool

File: 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#### Output
+- **Using**: Heroku
+- **Deliver**: live URLs to a `review` app, your `staging` & `production` apps
+- **With**: *independently*
+
+[Review app](https://fouille-pr-1.herokuapp.com)
+
+[Staging app](https://fouille-staging.herokuapp.com)
+
+[Production app](https://fouille.herokuapp.com)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md:10]: Live URLs. Good

File: 06-building-a-program-start-to-finish/setting-up-your-repository/output/README.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#### Output
+- **Using**: GitHub
+- **Deliver**: the repository for your proposed project and at least one `Pull Request`
+- **With**: *independently*
+
+[Repository](https://github.com/andela-pessien/inverted-index)
+
+[Pull Request](https://github.com/andela-pessien/inverted-index/pull/1)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-your-repository/output/README.md:8]: Cool

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+   * Creates an index for an array of books.
+   * @param {Array} books - The books to be indexed
+   * @param {String} filename - The name of the file the books come from
+   * @return {void}
+   */
+  createIndex(books, filename) {
+    this.filenames.push(filename);
+    const titles = [];
+    const index = {};
+    for (let i = 0; i < books.length; i += 1) {
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Whats your basis for using a for loop here when iterating through an Array?

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+      titles.push(books[i].title);
+      const bookTokens =
+        InvertedIndex.tokenize(`${books[i].title} ${books[i].text}`);
+      for (let j = 0; j < bookTokens.length; j += 1) {
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Whats your basis for using a for loop here when iterating through an Array of tokens?

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        }
+      }
+    }
+    this.titles.push(titles);
+    this.indices.push(index);
+  }
+
+  /**
+   * Removes the index for the books in a specified file.
+   * @param {String} filename - The name of the file to be removed
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Please use camelCase for you variables/ parameter names to make them more readable. Please address other instances of this.

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+   * Checks passed data for conformity to proper structure: an array of object
+   * literals
+   * each having a "title" and a "text" property.
+   * @param {Array} fileData - Data to be validated
+   * @returns {boolean} - Valid (true) or invalid (false)
+   * @static
+   */
+  static validateFile(fileData) {
+    if (Array.isArray(fileData) && fileData.length > 0) {
+      for (let i = 0; i < fileData.length; i += 1) {
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Whats your basis for using a for loop here when iterating through an Array?

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+      }
+
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Parses a passed string into unique tokens (individual words).
+   * @param {String} str - Sttring to be tokenized
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: What is "str"? Avoid single letter names or abbreviations. Be descriptive with your naming

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
+const InvertedIndex = require('../../src/inverted-index.js');
+const samples = require('../samples');
+const FileAPI = require('file-api');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: A base filename should exactly match the name of its default export. [See link](https://github.com/airbnb/javascript#naming--filename-matches-export)

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('should return true only if book data is properly structured', () => {
+      expect(InvertedIndex.validateFile(samples.validBooks)).toBe(true);
+    });
+  });
+
+  describe('Tokenization: ', () => {
+    it('should return the correct tokens for a given string', () => {
+      expect(InvertedIndex.tokenize(samples.validBooks[0].text)).toEqual(samples.tokens);
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: Can we have a test to strip off special characters and whitespace?

**Review by andela-moe — COMMENTED** [review]
Please review comments.

File: 06-building-a-program-start-to-finish/automating-tasks/output/README.md (hunk 1 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/automating-tasks/output/README.md:6]: Thank you!

File: 06-building-a-program-start-to-finish/checking-repository-health/output/README.md (hunk 2 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/checking-repository-health/output/README.md:6]: Thank you!

File: 06-building-a-program-start-to-finish/getting-peer-reviews/output/README.md (hunk 3 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/getting-peer-reviews/output/README.md:6]: Yes I was (most of it) especially the commit message convention and the `max-len` rule

File: 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md (hunk 4 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md:6]: Thank you!

File: 06-building-a-program-start-to-finish/setting-up-linters/output/README.md (hunk 5 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/setting-up-linters/output/README.md:11]: Thank you!

File: 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md (hunk 6 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md:10]: Thank you!

File: 06-building-a-program-start-to-finish/setting-up-your-repository/output/README.md (hunk 7 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/setting-up-your-repository/output/README.md:8]: Thank you!

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 8 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: I've changed it to a `forEach` loop, thanks

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 9 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: I've changed it to a `forEach` loop, thanks

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 10 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: I've fixed those, thank you

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 11 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: I've changed it to a `forEach` loop, thanks

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 12 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: I've changed that variable name, thanks

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js (hunk 13 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: I've fixed that, thank you

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js (hunk 14 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: I've added the test, thank you

File: 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    //   const bookTokens =
+    //     InvertedIndex.tokenize(`${books[i].title} ${books[i].text}`);
+    //   for (let j = 0; j < bookTokens.length; j += 1) {
+    //     if (bookTokens[j] in index) {
+    //       index[bookTokens[j]].push(books[i].title);
+    //     } else {
+    //       index[bookTokens[j]] = [books[i].title];
+    //     }
+    //   }
+    // }
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js]: Why do you have commented code here?

File: 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+   * @static
+   */
+  static tokenize(text) {
+    return [...new Set(text.toLowerCase()
+      .replace(/[^\s\w]|_/g, '')
+      .split(/\s+/))];
+  }
+}
+
+module.exports = InvertedIndex;
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js:159]: Good work

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 17]
```diff
@@ -0,0 +1,89 @@
+const InvertedIndex = require('../../src/InvertedIndex.js');
+const samples = require('../samples');
+const FileAPI = require('file-api');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: I believe your "FileAPI" variable should be in camelCase. This should also apply on your InvertedIndex class

**Review by andela-moe — COMMENTED** [review]
Please review comments

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js (hunk 17 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: Fixed, thanks

File: 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js (hunk 15 above)
Comment (andela-pessien) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js]: Oops, I've taken that out

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+const File = fileAPI.File;
+
+describe('Inverted Index Test Suite: ', () => {
+  describe('Reading file data: ', () => {
+    it('should read and return the contents of a file via callback', (done) => {
+      const file = new File('./validBooks.json');
+      InvertedIndex.readFile(file, (event) => {
+        expect(JSON.parse(event.target.result)).toEqual(samples.validBooks);
+        done();
+      })
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: I believe you are missing a semi-colon here

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+      this.invertedIndex.createIndex(samples.validBooks, 'validBooks.json');
+      this.invertedIndex.createIndex(samples.extraBooks, 'extraBooks.json');
+    });
+
+    it('should return books containing the specified keywords', () => {
+      expect(this.invertedIndex.searchIndex('the death he comes extravaganza', ['extraBooks.json', 'validBooks.json']).results)
+        .toEqual(samples.searchResults);
+    });
+  });
+});
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: You have a maximum line length of 80 violations on this file. Please address them all. 

One thing that can help you here is to ensure that your test use "Ubiquitous Language,"

It is a body of knowledge that can be understood by both the developer and the customer. This ubiquitous language is used to shape and develop the requirements and testing needed, at the level of the customer's understanding.

```
describe("Calculator: add", function (){
    it("should be able to add 2 numbers together", function (){
        var total = add(1, 2);
        expect(total).toBe(3);
    });

    it("should be able to add 3 numbers together", function (){
        var total = add(1, 2, 3);
        expect(total).toBe(6);
    });
});
```

If that don't work, then you'll want to use Template literals

**Review by andela-moe — COMMENTED** [review]
Please review comments.

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+      this.invertedIndex.createIndex(samples.extraBooks, 'extraBooks.json');
+    });
+
+    it('should return books containing the specified keywords', () => {
+      expect(this.invertedIndex.searchIndex('the death he comes extravaganza',
+      ['extraBooks.json', 'validBooks.json']).results)
+        .toEqual(samples.searchResults);
+    });
+  });
+});
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: You should also test that your InvertedIndex class has your defined functions/methods. If you are not sure of what this means, please ask your peers.

**Review by andela-moe — COMMENTED** [review]
Looking forward to the rest of your outputs

File: 06-building-a-program-start-to-finish/delivering-your-product/output/README.md [hunk 21]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: Google Doc
+- **Deliver**: `Release notes` for your final product in your repo's wiki
+- **With**: *independently*
+
+[Release Notes](https://github.com/andela-pessien/inverted-index/wiki/Release-Notes-(v1.0))
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/delivering-your-product/output/README.md:6]: Impressive

File: 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md [hunk 22]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: Medium
+- **Deliver**: a post reflecting on your learning experience while producing these outputs
+- **With**: *independently*
+
+[Reflection on learning experience](https://medium.com/@pj.essien/honey-i-broke-the-app-e4911d1277d1)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md:6]: Interesting read. Nice

File: 06-building-a-program-start-to-finish/using-design-patterns/output/README.md [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+#### Design Patterns
+
+##### Creational Pattern: Constructor Pattern
+In object-oriented programming a constructor is a special method used to initialize a newly created object once memory has been allocated for it. The `InvertedIndex` class uses a basic constructor pattern, which:
+- does not separate construction and representation
+- does not defer instantiation (i.e. specifies the class of object to be created)
+- allows for creation of more than one instance of a class
+The `InvertedIndex` class can be instantiated by calling the constructor:
+```const invertedIndex = new InvertedIndex();```
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/using-design-patterns/output/README.md:28]: Cool

File: 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js (hunk 20 above)
Comment (andela-moe) [reply 06-building-a-program-start-to-finish/writing-tests/output/test/spec/inverted-index-spec.js]: Cool

**Review by andela-moe — APPROVED** [review]
All good. Please schedule your defense.

**Review by andela-moe — APPROVED** [review]
Good work

**Pull request merged**